libs/zedUpload/azureutil/azure.go

```diff
@@ -311,9 +323,14 @@ func DownloadAzureBlob(accountURL, accountName, accountKey, containerName, remot
 					progressReceiver(progress)
 					progressLock.Unlock()
 				})
-			_, err = io.Copy(newSectionWriter(file, chunkStart+offset, count-offset, part), body)
-			body.Close()
-			return err
+			bp := *bufPool.Get().(*[]byte)
+			_, err = io.CopyBuffer(newSectionWriter(file, chunkStart+offset, count-offset, part), body, bp)
```

Comment:
I don't get it. What does this buy us? Changing from `io.Copy()` to `io.CopyBuffer()` just lets you allocate the single copy buffer, instead of go doing it. Go's default is 32*1024 [here](https://cs.opensource.google/go/go/+/refs/tags/go1.18.2:src/io/io.go;l=415), which is the same as we are doing, so what do we gain?

Reply:
`io.Copy()` also reuses the same buffer, see [here](https://cs.opensource.google/go/go/+/refs/tags/go1.18.2:src/io/io.go;l=425-451). It does not malloc and free for each 32KB copy, but uses the same one.

Reply:
Yes, for block we defined (i.e. 1MB), but not across blocks. So we will have thousands of allocations for large files.

Reply:
> for large files

Is that the use case? Where you do chunk-based download with large parallelism, so each chunk gets its own buffer allocated by `io.Copy()`, but if a large file file, say 100MB, needs to be downloaded, and you [defined chunk size as one MB](https://github.com/lf-edge/eve/pull/2643/files#diff-972a091cde07e2e7ac2e91d7269fc3a70508e17a19729d2cf4a1bbe187dd533aR281), then it is 100 chunks, each of which gets its own buffer allocated and freed?

Since you [defined parallelism as 16](https://github.com/lf-edge/eve/pull/2643/files#diff-972a091cde07e2e7ac2e91d7269fc3a70508e17a19729d2cf4a1bbe187dd533aR29), there will be at most 16 parallel downloads.

Those 16 will each have the allocation overhead (slightly larger with `sync.Pool` than with a straight `make([]byte, 32*1024)`, with the extra func call, but not terribly so).

Using the 100MB file, with 16 parallel downloads, we have about 6 reuses, as each parallel thread now can use an existing byte buffer from the pool.

OK, I get it now. Thanks for explaining. It probably is worth explaining it in the code as well, so people who come to it understand what we did and why (before someone tries to "simplify it" in the future).